### ISSUE TITLE
22

### ISSUE THREAD

**Issue by SteveHong1901** [SteveHong1901/STAT0023-Coursework-1#6]
`

import numpy as np
import scipy.stats as stats

class MulticlassLogisticRegression:
    # ... [Rest of your previous code]

    def _calculate_overdispersion(self, X_bias, y_encoded, probabilities):
        residuals = y_encoded - probabilities
        deviance = -2 * (np.sum(y_encoded * np.log(probabilities)) + np.sum((1 - y_encoded) * np.log(1 - probabilities)))
        degrees_freedom = X_bias.shape[0] - X_bias.shape[1]
        return deviance / degrees_freedom

    def hypothesis_test(self):
        X_bias = np.insert(self.X, 0, 1, axis=1)
        logits = X_bias.dot(self.weights)
        probabilities = self._softmax(logits)
        y_encoded = np.zeros((self.y.shape[0], len(self.classes)))
        for i, val in enumerate(self.y):
            y_encoded[i, val] = 1
        
        overdispersion = self._calculate_overdispersion(X_bias, y_encoded, probabilities)

        # Calculate variance-covariance matrix
        W = np.diag((probabilities * (1 - probabilities)).sum(axis=1))
        cov_matrix = np.linalg.inv(X_bias.T.dot(W).dot(X_bias))
        std_errors = np.sqrt(np.diagonal(cov_matrix) * overdispersion)

        # Conduct z-test for each coefficient
        z_scores = self.weights / std_errors[:, np.newaxis]
        p_values = 2 * (1 - stats.norm.cdf(np.abs(z_scores)))

        return z_scores, p_values

# Usage
# X_train, y_train = ...  # Your training data
# model = MulticlassLogisticRegression()
# model.X, model.y = X_train, y_train  # Store X and y in the model for hypothesis test
# model.fit(X_train, y_train)
# z_scores, p_values = model.hypothesis_test()
# print("Z-scores:", z_scores)
# print("P-values:", p_values)

import numpy as np
import scipy.stats as stats
import pandas as pd

class MulticlassLogisticRegression:
    # ... [Rest of your previous code]

    def hypothesis_test(self):
        X_bias = np.insert(self.X, 0, 1, axis=1)
        logits = X_bias.dot(self.weights)
        probabilities = self._softmax(logits)
        y_encoded = np.zeros((self.y.shape[0], len(self.classes)))
        for i, val in enumerate(self.y):
            y_encoded[i, val] = 1
        
        overdispersion = self._calculate_overdispersion(X_bias, y_encoded, probabilities)

        # Calculate variance-covariance matrix
        W = np.diag((probabilities * (1 - probabilities)).sum(axis=1))
        cov_matrix = np.linalg.inv(X_bias.T.dot(W).dot(X_bias))
        std_errors = np.sqrt(np.diagonal(cov_matrix) * overdispersion)

        # Conduct z-test for each coefficient
        z_scores = self.weights / std_errors[:, np.newaxis]
        p_values = 2 * (1 - stats.norm.cdf(np.abs(z_scores)))

        # Create table using pandas DataFrame
        variable_names = ['Intercept'] + [f'Feature_{i}' for i in range(1, self.X.shape[1] + 1)]
        table = pd.DataFrame({
            'Variable Name': np.repeat(variable_names, len(self.classes)),
            'Class': list(self.classes) * self.X.shape[1],
            'Weight': self.weights.flatten(),
            'Z-Score': z_scores.flatten(),
            'P-Value': p_values.flatten()
        })

        return table

# Usage
# X_train, y_train = ...  # Your training data
# model = MulticlassLogisticRegression()
# model.X, model.y = X_train, y_train  # Store X and y in the model for hypothesis test
# model.fit(X_train, y_train)
# results_table = model.hypothesis_test()
# print(results_table)

`
